server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java

```diff
@@ -0,0 +1,905 @@
+            try {
+                Matcher matcher = EXTRACT_VERSION.matcher(message);
+                matcher.find();
+                return new Version(Long.parseLong(matcher.group(2)), Long.parseLong(matcher.group(1)));
+            } catch (RuntimeException e) {
+                throw new RuntimeException("Unable to parse message: " + message, e);
+            }
+        }
+
+        @Override
+        public Optional<State> nextState(State currentState, Version input) {
+            return currentState.casFail(input, outputVersion);
+        }
+
+        @Override
+        public Version getVersion() {
+            return outputVersion;
+        }
+
+        @Override
+        public String toString() {
+            return "CASFail{" + outputVersion + "}";
+        }
+
+        @Override
+        public String getWriteableName() {
+            return "casfail";
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            outputVersion.writeTo(out);
+        }
+    }
+
+    private static class FailureHistoryOutput implements HistoryOutput {
+
+        private FailureHistoryOutput() {
+
+        }
+        private FailureHistoryOutput(@SuppressWarnings("unused") StreamInput streamInput) {
+
+        }
+
+        @Override
+        public Optional<State> nextState(State currentState, Version input) {
+            return currentState.fail();
+        }
+
+        @Override
+        public Version getVersion() {
+            return null;
+        }
+
+        @Override
+        public String toString() {
+            return "Fail";
+        }
+
+        @Override
+        public String getWriteableName() {
+            return "fail";
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            // nothing to write.
+        }
+    }
+
+    private static Function<Object, Object> missingResponseGenerator() {
+        return input -> new FailureHistoryOutput();
+    }
+
+    @SuppressForbidden(reason = "we only serialize data in the test case to be able to deserialize it for debugging")
+    private String base64Serialize(LinearizabilityChecker.History history) {
+        BytesStreamOutput output = new BytesStreamOutput();
+        try {
+            List<LinearizabilityChecker.Event> events = history.copyEvents();
+            output.writeInt(events.size());
+            for (LinearizabilityChecker.Event event : events) {
+                writeEvent(event, output);
+            }
+            output.close();
+            return Base64.getEncoder().encodeToString(BytesReference.toBytes(output.bytes()));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        } catch (ClassCastException e) {
+            return "Unable to produce base64 serialized version of history: " + history +"\n  Message: " + e.getMessage();
```

Comment:
should we fail in this case? This is unexpected, no?

Should we always check that serialization is not broken by always serializing histories, even if they have been linearizable?

Reply:
Fixed in 2f576d99d0d14bf375edc2158ce0a62210dfbe5e